lib/koala/api/graph_batch_api.rb

```diff
@@ -41,67 +41,106 @@ def graph_call_in_batch(path, args = {}, verb = "get", options = {}, &post_proce
-                raw_result = error
-              else
-                # (see note in regular api method about JSON parsing)
-                body = JSON.load("[#{call_result['body'].to_s}]")[0]
-
-                # Get the HTTP component they want
-                raw_result = case batch_op.http_options[:http_component]
-                when :status
-                  call_result["code"].to_i
-                when :headers
-                  # facebook returns the headers as an array of k/v pairs, but we want a regular hash
-                  parsed_headers
-                else
-                  body
-                end
-              end
-            end
-
-            # turn any results that are pageable into GraphCollections
-            # and pass to post-processing callback if given
-            result = GraphCollection.evaluate(raw_result, @original_api)
-            if batch_op.post_processing
-              batch_op.post_processing.call(result)
-            else
-              result
-            end
+      def generate_results
+        index = 0
+        lambda do |call_result|
+          batch_op     = batch_calls[index]; index += 1
+          post_process = batch_op.post_processing
+
+          # turn any results that are pageable into GraphCollections
+          result = GraphCollection.evaluate(
+            result_from_response(call_result, batch_op),
+            original_api
+          )
+
+          # and pass to post-processing callback if given
+          if post_process
+            post_process.call(result)
+          else
+            result
           end
         end
       end
 
+      def bad_response
+        # Facebook sometimes reportedly returns an empty body at times
+        BadFacebookResponse.new(200, '', 'Facebook returned an empty body')
+      end
+
+      def result_from_response(response, options)
+        return nil if response.nil?
+
+        headers   = coerced_headers_from_response(response)
+        error     = error_from_response(response, headers)
+        component = options.http_options[:http_component]
+
+        error || result_from_component({
+          :component => component,
+          :response  => response,
+          :headers   => headers
+        })
+      end
+
+      def coerced_headers_from_response(response)
+        headers = response.fetch('headers', [])
+
+        headers.each_with_object({}) do |h, memo|
+          memo.merge! h.fetch('name') => h.fetch('value')
+        end
+      end
+
+      def error_from_response(response, headers)
+        code = response['code']
+        body = response['body'].to_s
+
+        check_response(code, body, headers)
+      end
+
+      def batch_args
+        calls = batch_calls.map do |batch_op|
+          batch_op.to_batch_params(access_token, app_secret)
+        end
+
+        JSON.dump calls
+      end
+
+      def json_body(response)
+        body = response.fetch('body')
```

Comment:
This is another place where there's a new `KeyError` that wouldn't have occurred previously -- could we put back the `to_s`?